test/gui/swap.c

```diff
@@ -0,0 +1,431 @@
+/**
+ * @file
+ * Test code for mutt_window_swap()
+ *
+ * @authors
+ * Copyright (C) 2024 Dennis Schön <mail@dennis-schoen.de>
+ *
+ * @copyright
+ * This program is free software: you can redistribute it and/or modify it under
+ * the terms of the GNU General Public License as published by the Free Software
+ * Foundation, either version 2 of the License, or (at your option) any later
+ * version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "config/number.h"
+#define TEST_NO_MAIN
+#include "config.h"
+#include "acutest.h"
+#include "mutt/lib.h"
+#include "gui/lib.h"
+#include "test_common.h"
+
+// clang-format off
+static const char * initial_order[] = {
+  "apple",
+  "banana",
+  "cherry",
+  "damson",
+  "endive",
+  "fig",
+  "guava",
+  "hawthorn"
+};
+// clang-format on
+
+static const int num_children = mutt_array_size(initial_order);
+
+static void wdata_free(struct MuttWindow *win, void **ptr)
+{
+  if (!ptr || !*ptr)
+    return;
+
+  FREE(ptr);
+}
+
+static struct MuttWindow *new_window(const char *name)
+{
+  struct MuttWindow *win = mutt_window_new(WT_CUSTOM, MUTT_WIN_ORIENT_VERTICAL,
+                                           MUTT_WIN_SIZE_MAXIMISE, MUTT_WIN_SIZE_UNLIMITED,
+                                           MUTT_WIN_SIZE_UNLIMITED);
+  win->wdata = mutt_str_dup(name);
+  win->wdata_free = wdata_free;
+
+  return win;
+}
+
+static struct MuttWindow *new_parent(void)
+{
+  struct MuttWindow *parent = new_window("parent");
+
+  for (int i = 0; i < num_children; i++)
+    mutt_window_add_child(parent, new_window(initial_order[i]));
+
+  return parent;
+}
+
+static struct MuttWindow *find_child(struct MuttWindow *parent, const char *name)
+{
+  struct MuttWindow *win = NULL;
+  TAILQ_FOREACH(win, &parent->children, entries)
+  {
+    if (mutt_str_equal(win->wdata, name))
+      return win;
+  }
+
+  ASSERT(false);
+}
+
+static void check_order(struct MuttWindow *parent, const char **expected)
```

Comment:
All the removing/inserting is happening through the `TAILQ_` functions, so it ought to be safe,
but (if we were being paranoid) this function only checks that the **forward** links are correct
(`TAILQ` is doubly linked)

Reply:
I also think that would be pretty paranoid. If you really mess with the internal structure like that, you're on your own. 😆